helix-core/src/main/java/org/apache/helix/controller/stages/CurrentStateComputationStage.java

```diff
@@ -129,7 +148,14 @@ private void updatePendingMessages(LiveInstance instance, Collection<Message> pe
         String partitionName = message.getPartitionName();
         Partition partition = resource.getPartition(partitionName);
         if (partition != null) {
-          setMessageState(currentStateOutput, resourceName, partition, instanceName, message);
+          String currentState = currentStateOutput.getCurrentState(resourceName, partition,
+              instanceName);
+          if (_isTaskFrameworkPipeline || !isStaleMessage(message, currentState)) {
+            setMessageState(currentStateOutput, resourceName, partition, instanceName, message);
+          } else {
+            existingStaleMessages.putIfAbsent(instanceName, new HashMap<>());
```

Comment:
As mentioned above, I prefer this to be an explicit set method, addStaleMessage()